### PR TITLE
Add @cover annotation in test-load file and add support for codecov

### PR DESCRIPTION
**Pull request by sarthak-19**
## Summary
This PR adds `@cover` annotation for test cases and add .yml file that can be used to configure codecov for automated test coverage in github actions.

Addresses https://github.com/WordPress/performance/issues/1284

## Relevant technical choices

I've addressed the following file `plugins/webp-uploads/tests/test-load.php`  for adding `@cover` annotation, if the approach is approved will extend this to other files.

### CODE REVIEW

File: composer.json
Comment (swissspidy) [inline composer.json]: This isn't needed, PHPUnit is installed via `yoast/phpunit-polyfills` above

File: composer.json
Comment (swissspidy) [inline composer.json]: This isn't needed either, nor does it actually exist

File: .github/workflows/codecov-coverage.yml
Comment (swissspidy) [inline .github/workflows/codecov-coverage.yml]: Can we run coverage in the existing `php-test-plugins.yml` workflow instead?

**Comment by swissspidy** [comment]
@sarthak-19 Welcome, and thanks for opening this! I just stumbled upon the PR and left a couple of comments for things that caught my eye.

**Comment by sarthak-19** [comment]
> @sarthak-19 Welcome, and thanks for opening this! I just stumbled upon the PR and left a couple of comments for things that caught my eye.

@swissspidy Thank you so much for the feedback. I've addressed those and looking forward to keep improving 😄

File: .github/workflows/php-test-plugins.yml
Comment (westonruter) [inline .github/workflows/php-test-plugins.yml]: Since collecting code coverage takes much longer, as I recall, should the coverage only happen for one of the jobs instead of for all of them?

Comment (sarthak-19) [reply .github/workflows/php-test-plugins.yml]: Yes you're right, collecting code coverage takes time, but I'm not quite sure that for which job it'll be ideal to run.
Should I run for `PHP v8.0` and if this is a viable option, then will the below implementation work since we are using the latest WP version anyways?
```yml

      - name: Running single site unit tests with coverage
        if: matrix.php == '8.0'
        run: npm run test-php -- --coverage-clover=coverage.xml
      - name: Running multisite unit tests with coverage
        if: matrix.php == '8.0'
        run: npm run test-php-multisite -- --coverage-clover=coverage-multisite.xml

      - name: Upload single site coverage reports to Codecov
        if: matrix.php == '8.0'
        uses: codecov/codecov-action@v4
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: coverage.xml
          flags: unittests
          name: ${{ matrix.php }}-single-site-coverage
      - name: Upload multisite coverage reports to Codecov
        if: matrix.php == '8.0'
        uses: codecov/codecov-action@v2
        with:
          token: ${{ secrets.CODECOV_TOKEN }}
          files: coverage-multisite.xml
          flags: multisite
          name: ${{ matrix.php }}-multisite-coverage
```

cc : @westonruter

**Comment by pbearne** [comment]
I am really glad to see I have been try to get into core ten+ years

File: .github/workflows/php-test-plugins.yml
Comment (westonruter) [reply .github/workflows/php-test-plugins.yml]: You may want to refer to what was done for the AMP plugin: https://github.com/ampproject/amp-wp/blob/60a655cdad9e3431b47b2dcc3b2e5ddb86ac5c5e/.github/workflows/build-test-measure.yml#L261-L337

Note the new `coverage` matrix variable which is only set to `true` for one configuration. This would be better than having to test against a specific PHP version everywhere. Yes, I think running coverage on the most recent PHP version would be best as it should complete the fastest. So maybe do it in PHP 8.2?

cc @thelovekesh

**Comment by sarthak-19** [comment]
> Note the new coverage matrix variable which is only set to true for one configuration. This would be better than having to test against a specific PHP version everywhere. Yes, I think running coverage on the most recent PHP version would be best as it should complete the fastest. So maybe do it in PHP 8.2?

Apologies for the delayed response, made the suggested changes.
Does this works or any other changes are to be made, highly appreciate your insights here.

cc : @westonruter

**Comment by westonruter** [comment]
@sarthak-19 That looks right!

File: .github/workflows/php-test-plugins.yml
Comment (westonruter) [inline .github/workflows/php-test-plugins.yml:103]: Is `multisite` a valid flag?  Shouldn't this be `unittests`?

```suggestion
          flags: unittests
```

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml:54]: I wonder, given we include this as an extra scenario, should we maybe remove `8.2` from the regular `php` list above? Otherwise we'd be running the exact same test (WP latest, PHP 8.2) twice, only with and without code coverage, which shouldn't change the outcome of the tests.

If there's any benefit to keeping both (which would only be because of the implications that running code coverage could have), then I think the better choice would be to implement a separate workflow entirely.

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml]: Maybe append commit SHA to avoid chance of conflicts?
```suggestion
            npm run test-php -- --coverage-clover=coverage-${{ github.sha }}.xml
```

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml]: See above:
```suggestion
            npm run test-php-multisite -- --coverage-clover=coverage-multisite-${{ github.sha }}.xml
```

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml]: See above:
```suggestion
          files: coverage-${{ github.sha }}.xml
```

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml]: See above:
```suggestion
          files: coverage-multisite-${{ github.sha }}.xml
```

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml]: How about we use `single` instead of `unittests`, since the differentiator between the two datasets is "single site" vs "multisite", not "unit tests" vs "multisite".
```suggestion
          flags: single
```

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [reply .github/workflows/php-test-plugins.yml:103]: These are arbitrary values to separate test scenarios, so I think passing `multisite` makes sense here (separating from "single site").

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml:95]: Let's make sure we're alerted of failures for coverage reporting:
```suggestion
          name: ${{ matrix.php }}-single-site-coverage
          fail_ci_if_error: true
```

File: .github/workflows/php-test-plugins.yml
Comment (felixarntz) [inline .github/workflows/php-test-plugins.yml:104]: See above:
```suggestion
          name: ${{ matrix.php }}-multisite-coverage
          fail_ci_if_error: true
```

**Review by felixarntz — COMMENTED** [review]
@sarthak-19 Thank you for the PR! This looks quite solid, a few suggestions and questions.

File: .github/workflows/php-test-plugins.yml
Comment (westonruter) [reply .github/workflows/php-test-plugins.yml]: How would there be a conflict? Each GHA job should run from a pristine state, no? For the `coverage.xml` file to persist across runs this would require some explicit caching.

Comment (felixarntz) [reply .github/workflows/php-test-plugins.yml]: Probably. I find this safer, but either works for me.

File: .github/workflows/php-test-plugins.yml
Comment (sarthak-19) [reply .github/workflows/php-test-plugins.yml:54]: Yes make sense, removing the 8.2 version from PHP list.

File: .github/workflows/php-test-plugins.yml
Comment (sarthak-19) [reply .github/workflows/php-test-plugins.yml]: > I find this safer.

I think better safe than sorry.
So implementing this as well. 😀

**Comment by westonruter** [comment]
FYI: I've added the `CODECOV_TOKEN` to the repository secrets.

**Review by westonruter — APPROVED** [review]
I suppose the one job will continue to fail since this PR was opened from a fork. I'll leave it to Felix to approve and merge.

**Review by felixarntz — APPROVED** [review]
Thanks for the PR @sarthak-19, LGTM!

Before we merge though: Doesn't one of us has to configure / set up this project on https://about.codecov.io/ first? 🤔

**Comment by westonruter** [comment]
@felixarntz I _believe_ it's all already configured according to the steps listed at https://app.codecov.io/gh/WordPress/performance

The only thing outstanding is the third step, which is to merge into the preferred branch:

> image

I'll do so now!

**Comment by westonruter** [comment]
Alas, no: https://github.com/WordPress/performance/actions/runs/11823769472/job/32943781732

**Comment by westonruter** [comment]
I've been chatting with @thelovekesh about what's going on. 

For one, we're missing some additional `--` arg separators since we're passing from `npm` to `composer` and then to `phpunit`. So here:

https://github.com/WordPress/performance/blob/745c145709e8d6eec655c109f472ca77b151178d/.github/workflows/php-test-plugins.yml#L77

https://github.com/WordPress/performance/blob/745c145709e8d6eec655c109f472ca77b151178d/.github/workflows/php-test-plugins.yml#L84

These need to instead be, respectively:

```bash
npm run test-php -- -- --coverage-clover=coverage-${{ github.sha }}.xml 
```

and 

```bash
npm run test-php-multisite -- -- --coverage-clover=coverage-multisite-${{ github.sha }}.xml
```

Additionally, the coverage report for each plugin is getting overwritten to the same XML file. There needs to be separate XML files generated for each plugin.

Lovekesh has some additional thoughts.

**Comment by westonruter** [comment]
I've opened https://github.com/WordPress/performance/pull/1652 to disable Codecov for the time being.

**Pull request merged**
